providers/repo.rb

```diff
@@ -138,17 +138,45 @@ def install_gem
   end
 end
 
-def read_token(repo_url, gems=false)
-  return repo_url unless new_resource.master_token
+### The hostname lookup chain is:
+### 1) node['packagecloud']['override_hostname']
+### 2) node['fqdn']
+### 3) node['hostname']
+### We persist this hostname result in node['packagecloud']['_stored_hostname']
+### between chef runs so that we can determine if a new read token should be issued
+### from the packagecloud API for new/changed hostnames. See #should_issue_read_token?
+def get_hostname
+  hostname = override_or_fqdn_or_hostname
+  if hostname.nil?
+    raise("Can't determine hostname!  Set node['packagecloud']['override_hostname'] " \
+          "if it cannot be automatically determined by Ohai.")
+  else
+    node.set['packagecloud']['_stored_hostname'] = hostname
+    hostname
+  end
+end
 
-  if node['fqdn'].nil?
-    Chef::Log.fatal("This node's fqdn is set to nil, so a read token cannot be issued!" \
-                    "Please change your fqdn settings.")
+## Legacy deployments who have node['packagecloud']['hostname'] set are
+## migrated over
+def stored_hostname
+  if node['packagecloud']['hostname']
+    node.set['packagecloud']['_stored_hostname'] = node['packagecloud']['hostname']
   end
+  node['packagecloud']['_stored_hostname']
+end
 
-  if node['packagecloud'][new_resource.repository].nil? ||
-     node['packagecloud']['hostname'] != node['fqdn']
+def should_issue_read_token?
```

Comment:
You could monkeypatch the providers, but that's probably not a great idea.  

I don't know, it really seems to me like handling token generation and validation should be PackageCloud's burden and read tokens shouldn't be cached stored client-side.  Otherwise you can't generate temporal tokens, and I just see a big security risk in centralizing their storage in an unencrypted database.  Don't get me wrong; I am a big fan of caching, but in this case, I don't find it persuasive.